postings.go

```diff
@@ -158,24 +154,31 @@ func Merge(its ...Postings) Postings {
 }
 
 type mergePostings struct {
-	a, b     Postings
-	aok, bok bool
-	cur      uint32
+	a, b        Postings
+	initialized bool
+	aok, bok    bool
+	cur         uint32
 }
 
 func newMergePostings(a, b Postings) *mergePostings {
-	it := &mergePostings{a: a, b: b}
-	it.aok = it.a.Next()
-	it.bok = it.b.Next()
-
-	return it
+	return &mergePostings{a: a, b: b, aok: true, bok: true}
 }
 
 func (it *mergePostings) At() uint32 {
 	return it.cur
 }
 
 func (it *mergePostings) Next() bool {
+	if !it.initialized {
+		if !it.a.Next() {
+			it.aok = false
+		}
+		if !it.b.Next() {
+			it.bok = false
+		}
+		it.initialized = true
+	}
```

Comment:
Swapping function is likely going to be the slowest alternative.

I was referring to the usage of

```
if !it.a.Next() {
  it.aok = false
}
if !it.b.Next() {
  it.bok = false
}
```

Previously, in the constructor we just did:


```
  it.aok = it.a.Next()
  it.bok = it.b.Next()
```

Semantically seem equivalent. What's speaking against keeping it like that.

And yes, also wondering why the additional attribute instead of keeping the initialization in the constructor. I'm not sure I know which part of your other response you meant.